harness/config/external-images.yml

```diff
@@ -36,11 +36,27 @@ command('external-images config [--skip-exists]'):
     }
     echo \Symfony\Component\Yaml\Yaml::dump($compose, 100, 2);
 
-command('external-images pull'): |
-  #!bash(workspace:/)|@
-  CONF_ARGS=()
-  if [ -n "${CI:-}" ] || [ -n "${BUILD_ID:-}" ]; then
-    CONF_ARGS+=(--skip-exists)
-  fi
-
-  passthru "ws external-images config $(printf '%q ' "${CONF_ARGS[@]}") | docker-compose -f - pull"
+command('external-images pull'):
+  env:
+    EXTERNAL_REGISTRIES: = json_encode(@('docker.external_image_registries'))
+  exec: |
+    #!php
+    foreach (json_decode($env['EXTERNAL_REGISTRIES'], true) as $registry) {
+      $pass = escapeshellarg($registry['password']);
+      $user = escapeshellarg($registry['username']);
+      $url = escapeshellarg($registry['url']);
+      shell_exec("echo $pass | docker login --username=$user --password-stdin $url");
+    }
+
+    $confArgs = [];
+    if (strlen(@$env['CI']) > 0 || strlen(@$env['BUILD_ID']) > 0) {
+      $confArgs[] = "--skip-exists";
+    }
+    $confArgs = escapeshellarg(join(' ', $confArgs));
+
+    shell_exec("ws external-images config $confArgs | docker-compose -f - pull");
```

Comment:
passthru() would be useful here, but it's not like the bash function ws has, so best logging the command run before running it

Reply:
also php doesn't exception on errors here